cranelift/codegen/src/isa/x64/inst/emit.rs

```diff
@@ -2433,6 +2441,116 @@ pub(crate) fn emit(
+        Inst::LockCmpxchg {
+            ty,
+            src,
+            dst,
+            srcloc,
+        } => {
+            if let Some(srcloc) = srcloc {
+                sink.add_trap(*srcloc, TrapCode::HeapOutOfBounds);
+            }
+            // lock cmpxchg{b,w,l,q} %src, (dst)
+            // Note that 0xF0 is the Lock prefix.
+            let (prefix, rex, opcodes) = match *ty {
+                types::I8 => {
+                    let mut rex_flags = RexFlags::clear_w();
+                    let enc_src = int_reg_enc(*src);
+                    if enc_src >= 4 && enc_src <= 7 {
+                        rex_flags.always_emit();
+                    };
+                    (LegacyPrefixes::_F0, rex_flags, 0x0FB0)
+                }
+                types::I16 => (LegacyPrefixes::_66F0, RexFlags::clear_w(), 0x0FB1),
+                types::I32 => (LegacyPrefixes::_F0, RexFlags::clear_w(), 0x0FB1),
+                types::I64 => (LegacyPrefixes::_F0, RexFlags::set_w(), 0x0FB1),
+                _ => unreachable!(),
+            };
+            emit_std_reg_mem(sink, prefix, opcodes, 2, *src, &dst.finalize(state), rex);
+        }
+
+        Inst::AtomicRMW { ty, op, srcloc } => {
+            /* Emit this:
+
+                  mov{b,w,l,q}           (%r9), %rax  // rax = old value
+                 again:
+                  movq                   %rax, %r11   // rax = old value, r11 = old value
+                  `op`q                  %r10, %r11   // rax = old value, r11 = new value
+                  lock cmpxchg{b,w,l,q}  %r11, (%r9)  // try to store new value
+                  jnz again // If this is taken, rax will have a "revised" old value
+
+               Operand conventions:
+                  IN:  %r9 (addr), %r10 (2nd arg for `op`)
+                  OUT: %rax (old value), %r11 (trashed), %rflags (trashed)
+
+               In the case where the operation is 'xchg', the second insn is instead
+                 movq                    %r10, %r11
+               so that we simply write in the destination, the "2nd arg for `op`".
+            */
+            let rax = regs::rax();
+            let r9 = regs::r9();
+            let r10 = regs::r10();
+            let r11 = regs::r11();
+            let rax_w = Writable::from_reg(rax);
+            let r11_w = Writable::from_reg(r11);
+            let amode = Amode::imm_reg(0, r9);
+            let amode_rm = RegMem::mem(amode.clone());
+            let again_label = sink.get_label();
+
+            // mov{b,w,l,q} (%r9), %rax
+            // No need to call `add_trap` here, since the `i1` emit will do that.
+            let i1 = match *ty {
+                types::I8 => Inst::movzx_rm_r(ExtMode::BQ, amode_rm, rax_w, *srcloc),
+                types::I16 => Inst::movzx_rm_r(ExtMode::WQ, amode_rm, rax_w, *srcloc),
+                types::I32 => Inst::movzx_rm_r(ExtMode::LQ, amode_rm, rax_w, *srcloc),
+                types::I64 => Inst::mov64_rm_r(amode_rm, rax_w, *srcloc),
```

Comment:
With the new `load` helper introduced in the SIMD PR (that should land soonish, probably), you might be able to just use `Inst::load` here.

Reply:
Yeah, I just spotted that.

Reply:
Should be in `main` now.